yarn-project/pxe/src/pxe_service/pxe_service.ts

```diff
@@ -756,6 +760,61 @@ export class PXEService implements PXE {
     );
   }
 
+  // As above, this is a hack for encrypted/unencrypted logs ordering, now they are sorted. Since the private kernel
+  // cannot keep track of side effects that happen after or before a nested call, we override the gathered logs.
+  // As a sanity check, we at least verify that the elements are the same, so we are only tweaking their ordering.
+  // See yarn-project/end-to-end/src/e2e_ordering.test.ts
+  // See https://github.com/AztecProtocol/aztec-packages/issues/1641
+  // Added as part of resolving #5017
+  private patchLogsOrdering(execResult: ExecutionResult) {
+    const encLogs = collectEncryptedLogs(execResult).flatMap(l => l.logs);
+    const unencLogs = collectUnencryptedLogs(execResult).flatMap(l => l.logs);
+    const getLogs = (res: ExecutionResult, enc: boolean) => {
+      const logs: SideEffect[] = enc
+        ? res.callStackItem.publicInputs.encryptedLogsHashes.concat(res.nestedExecutions.flatMap(e => getLogs(e, true)))
+        : res.callStackItem.publicInputs.unencryptedLogsHashes.concat(
+            res.nestedExecutions.flatMap(e => getLogs(e, false)),
+          );
+
+      return logs;
+    };
+
+    const sortSEs = (a: SideEffect, b: SideEffect) => {
+      if (a.isEmpty()) {
+        return 1;
+      } else if (b.isEmpty()) {
+        return -1;
+      } else {
+        return Number(a.counter.toBigInt() - b.counter.toBigInt());
+      }
+    };
+
+    const sortedEncLogs = getLogs(execResult, true).sort(sortSEs);
+    const sortedUnencLogs = getLogs(execResult, false).sort(sortSEs);
+
+    const finalEncLogs: EncryptedL2Log[] = [];
+    sortedEncLogs.forEach((sideEffect: SideEffect) => {
+      if (!sideEffect.isEmpty()) {
+        const isLog = (log: EncryptedL2Log) => Fr.fromBuffer(log.hash()).equals(sideEffect.value);
+        const thisLogIndex = encLogs.findIndex(isLog);
+        finalEncLogs.push(encLogs[thisLogIndex]);
+      }
+    });
+
+    const finalUnencLogs: UnencryptedL2Log[] = [];
+    sortedUnencLogs.forEach((sideEffect: SideEffect) => {
+      if (!sideEffect.isEmpty()) {
+        const isLog = (log: UnencryptedL2Log) => Fr.fromBuffer(log.hash()).equals(sideEffect.value);
+        const thisLogIndex = unencLogs.findIndex(isLog);
+        finalUnencLogs.push(unencLogs[thisLogIndex]);
+      }
+    });
+
+    const encryptedLogs = new EncryptedTxL2Logs([new EncryptedFunctionL2Logs(finalEncLogs)]);
+    const unencryptedLogs = new UnencryptedTxL2Logs([new UnencryptedFunctionL2Logs(finalUnencLogs)]);
+    return { encryptedLogs, unencryptedLogs };
+  }
+
```

Comment:
This hack was added to mimic `patchPublicCallStackOrdering` for the same reasons - the tx in ts doesn't have the sorting information the circuit does, so we have to re-sort the ts logs.

Reply:
Seems to me it's not a hack but a necessary thing to do! Maybe we can provide an instance when running the simulator, and every time a log is emitted it's pushed to an array in this instance. Then at the end of the simulation we can just retrieve the logs from it that are already "sorted".

Reply:
I believe this happens already (`ClientExecutionContext`?) but nested executions push to a subarray which then gets bubbled up, so we need to re-sort after. Maybe once we hash inside the circuit, we could use the oracle call to push a log and counter outside, so when logs are bubbled up they could be sorted before this ^ call?

Reply:
Got it - will add in another PR alongside the hash change and tx.ts error fix!